File: 1439.find-the-kth-smallest-sum-of-a-matrix-with-sorted-rows.cs
/*
 * @lc app=leetcode id=1439 lang=csharp
 *
 * [1439] Find the Kth Smallest Sum of a Matrix With Sorted Rows
 */

// @lc code=start
public class Solution {
    public List<int> KSmallestPairs(int[] nums1, int[] nums2, int k)
    {
        PriorityQueue<Tuple<int, int>, int> heap = new PriorityQueue<Tuple<int, int>, int>();
        int n1 = nums1.Length;
        int n2 = nums2.Length;
        Tuple<int, int> newTuple;
        HashSet<Tuple<int, int>> visited = new HashSet<Tuple<int, int>>();
        List<int> result =  new List<int>();

        heap.Enqueue(new Tuple<int,int> (0, 0), nums1[0] + nums2[0]);
        visited.Add(new Tuple<int,int> (0, 0));


        while (result.Count < k && heap.TryDequeue(out Tuple<int, int> t, out int total))
        {
            newTuple = new Tuple<int,int> (t.Item1+1, t.Item2);
            if (t.Item1 < n1 - 1 && !visited.Contains(newTuple))
            {
                heap.Enqueue(newTuple, total - nums1[t.Item1] + nums1[t.Item1+1]);
                visited.Add(newTuple);
            }
            newTuple = new Tuple<int,int> (t.Item1, t.Item2+1);
            if (t.Item2 < n2 - 1 && !visited.Contains(newTuple))
            {
                heap.Enqueue(newTuple, total - nums2[t.Item2] + nums2[t.Item2+1]);
                visited.Add(newTuple);
            }
            result.Add(nums1[t.Item1] + nums2[t.Item2]);
        }
        return result;
    }
    public int KthSmallest(int[][] mat, int k) {
        int m = mat.Length;
        List<int> pairs = mat[0].ToList();

        for (int i = 0; i < m-1; i++)
        {
            pairs = KSmallestPairs(pairs.ToArray(), mat[i+1], 200);
        }
        return pairs[k - 1];
    }
}
// @lc code=end

File: 1439.find-the-kth-smallest-sum-of-a-matrix-with-sorted-rows.py
#
# @lc app=leetcode id=1439 lang=python3
#
# [1439] Find the Kth Smallest Sum of a Matrix With Sorted Rows
#

# @lc code=start
import heapq


class Solution:
    # k is rather small (<= 200)
    # Therefore we can refer to the solution given in Leetcode 373
    # Key idea: because all the elements contribute to the final sum, the min sum list of mat[:i+1] = min sum combinations between combinations of mat[:i] and mat[i]
    
    def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
        heap, n1, n2, visited = [], len(nums1), len(nums2), set()
        heap.append([nums1[0] + nums2[0], 0, 0])
        visited.add((0,0))

        result = []
        while heap and len(result) < k:
            total, i, j = heapq.heappop(heap)
            if i < n1-1 and (i+1, j) not in visited:
                heapq.heappush(heap, [total - nums1[i] + nums1[i+1], i+1, j])
                visited.add((i+1,j))
            if j < n2-1 and (i, j+1) not in visited:
                heapq.heappush(heap, [total - nums2[j] + nums2[j+1], i, j+1])
                visited.add((i,j+1))
            result.append(nums1[i] + nums2[j])
        return result
    
    
    
    def kthSmallest(self, mat: List[List[int]], k: int) -> int:
        m = len(mat)
        pairs = mat[0]
        for i in range(m-1):
            pairs = self.kSmallestPairs(pairs, mat[i+1], 200)
        return pairs[k-1]
        
# @lc code=end

